File: palindrome.erl
-module(palindrome).
-export([isPalindrome/1]).
-export([display/1]).
-export([checkIfCanBuildPalindrome/1]).
-export([buildPalindrome/2]).

isPalindrome(InputString) ->
  InputString == lists:reverse(InputString).

buildPalindrome(EvensMap, OddsMap) ->
  % Character duplication graciously borrowed from
  % http://stackoverflow.com/questions/12973566/list-of-repeating-element-in-erlang
  EvensStr = dict:fold(
    fun(Key, Count, Acc) ->
      string:concat(Acc, lists:flatten(lists:duplicate(Count div 2, Key)))
    end,
  "", EvensMap),
  OddsStr  = dict:fold(
    fun(Key, Count, Acc) when Count > 1 ->
      string:concat(Acc, lists:flatten(lists:duplicate(Count div 2, Key)));
      (Key, _, Acc) -> string:concat(Acc, [Key])
    end,
  "", OddsMap),
  string:concat(string:concat(EvensStr, OddsStr), lists:reverse(EvensStr)).

checkIfCanBuildPalindrome(InputString) ->
  UpdateFun = fun(Old)    -> Old + 1 end,
  FoldFun   = fun(Char, Map) -> dict:update(Char, UpdateFun, 1, Map) end,
  OccurrencesMap = lists:foldl(FoldFun, dict:new(), InputString),
  EvensMap = dict:filter(fun(_, Value) -> Value rem 2 == 0 end, OccurrencesMap),
  OddsMap  = dict:filter(fun(_, Value) -> Value rem 2 /= 0 end, OccurrencesMap),
  OddsSize = dict:size(OddsMap),
  if
    OddsSize > 1 ->
      io:format("~s cannot be made into a palindrome~n", [InputString]);
    true ->
      io:format("A palindromic form of ~s is ~s~n", [InputString, buildPalindrome(EvensMap, OddsMap)])
  end.

display(InputString) ->
  IsPalindrome = isPalindrome(InputString),
  if
    IsPalindrome ->
      io:format("~s is a palindrome!~n", [InputString]);
    true ->
      checkIfCanBuildPalindrome(InputString)
  end.

File: palindrome.scala
def isPalindrome(ls: List[Char]): Boolean = ls == ls.reverse

def canBePalindrome(ls: List[Char]): (Boolean, String) = {
  // Graciously borrowed from solution found on
  // http://stackoverflow.com/questions/11448685/
  //   scala-how-can-i-count-the-number-of-occurrences-in-a-list
  val map = ls.groupBy(identity).mapValues(_.size)

  map count ((entry: (Char, Int)) => entry._2 % 2 != 0) match {
    case n if n > 1 => (false, ls.mkString)
    case _          => {
      val oddsMap  = map filter ((entry: (Char, Int)) => entry._2 % 2 != 0)
      val evensMap = map filter ((entry: (Char, Int)) => entry._2 % 2 == 0)
      var buildStr = ""
      for (entry <- evensMap) {
        for (i <- 1 to entry._2 / 2) {
          buildStr += entry._1
        }
      }

      var single = '\u0000'
      for (entry <- oddsMap) {
        if (entry._2 > 1) {
          for (i <- 1 to entry._2 / 2)
            buildStr += entry._1
        } else {
          single = entry._1
        }
      }

      (true, buildStr + single + buildStr.reverse)
    }
  }
}


if (isPalindrome(args(0).toList)) println(args(0) + " is a palindrome")
else {
  val finalStr = canBePalindrome(args(0).toList) match {
    case (true, ls)  => ls.mkString
    case _           => "impossible"
  }

  println("A palindromic form of " + args(0) + " is " + finalStr)
}
